### PR TITLE
Ensure duplicate incoming blocks share the same incoming values

### PR DESCRIPTION
**Pull request by kuhar**
This is to fix the issue #208.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

**Comment by JacobHeAMD** [comment]
ok to test

**Comment by JacobHeAMD** [comment]
test this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit a50a637c37db1544f581094d97c75d224590af88
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev 1fc1a7d4248b4749c3df21eb48f7ae97b6cddf74</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

File: patch/llpcPatchPeepholeOpt.cpp [hunk 1]
```diff
@@ -582,6 +582,12 @@ void PatchPeepholeOpt::visitPHINode(
 
             pResult = InsertElementInst::Create(pResult, pNewPhiNode, pElementIndex, "", pInsertPos);
 
+            // Make sure the same incoming blocks have identical incoming values.
+            // If we have already inserted an incoming arc for a basic block,
+            // reuse the same value in the future incoming arcs from the same block.
+            SmallDenseMap<BasicBlock *, Value *, 8> incomingBlockToValue;
```
Comment (amdrexu) [inline patch/llpcPatchPeepholeOpt.cpp]: SmallDenseMap<BasicBlock *, Value *, 8> -> SmallDenseMap<BasicBlock*, Value*, 8>
BasicBlock * -> BasicBlock*
Value * -> Value*
We treat "Type*" as a whole entity.

incomingBlockToValue -> incomingPairMap

File: patch/llpcPatchPeepholeOpt.cpp [hunk 2]
```diff
@@ -591,17 +597,28 @@ void PatchPeepholeOpt::visitPHINode(
 
                 if (Instruction* const pInst = dyn_cast<Instruction>(pIncoming))
                 {
-                    ExtractElementInst* const pExtractElement = ExtractElementInst::Create(
-                        pIncoming, pElementIndex);
+                    Value *pNewIncomingValue = nullptr;
```
Comment (amdrexu) [inline patch/llpcPatchPeepholeOpt.cpp]: Value * -> Value*

**Comment by kuhar** [comment]
Thanks for the comments @amdrexu. Do you have a .clang-fornat file I can use to format this and future commits?

**Comment by amdrexu** [comment]
> Thanks for the comments @amdrexu. Do you have a .clang-fornat file I can use to format this and future commits?

No, we haven't because original development was not performed on Git. We are thinking about this. Currently, you can reference existing codes (except translator folder). We are also considering publishing our internal PAL coding style.

**Comment by amdvlk-admin** [comment]
### Test summary for commit a50a637c37db1544f581094d97c75d224590af88
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev 1fc1a7d4248b4749c3df21eb48f7ae97b6cddf74</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 8475c03cad8e7b0fbdbf66df322043ebe9a93d49
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev 1fc1a7d4248b4749c3df21eb48f7ae97b6cddf74</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 03f5c6078f700e3ace41e2f5fa1d918a0c49ae71
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev 1fc1a7d4248b4749c3df21eb48f7ae97b6cddf74</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**